File: src/main/java/com/vc/hard/L975.scala
package com.vc.hard

object L975 {
  def oddEvenJumps(arr: Array[Int]): Int = {
    val n = arr.length

    var i = n - 1
    val map = new java.util.TreeMap[Int, Int]()
    val lower = new Array[Boolean](n)
    val higher = new Array[Boolean](n)

    map.put(arr(i), i)
    lower(i) = true
    higher(i) = true

    var res = 1
    i -= 1
    while(i >= 0) {
      val lo = map.floorEntry(arr(i))
      val hi = map.ceilingEntry(arr(i))

      if(lo != null) lower(i) = higher(lo.getValue)
      if(hi != null) higher(i) = lower(hi.getValue)

      if(higher(i)) res += 1

      map.put(arr(i), i)
      i -= 1
    }
    res
  }
}

File: src/main/java/com/vc/hard/L975.java
package com.vc.hard;

import java.util.TreeMap;

class L975 {
    public int oddEvenJumps(int[] arr) {
        TreeMap<Integer, Integer> map = new TreeMap<>();
        int n = arr.length;
        boolean[] lower = new boolean[n];
        boolean[] higher = new boolean[n];

        lower[n - 1] = higher[n - 1] = true;
        map.put(arr[ n - 1], n - 1);

        int res = 1;
        for(int i = n - 2; i >= 0; i--){
            Integer lo = map.floorKey(arr[i]);
            Integer hi = map.ceilingKey(arr[i]);

            if(lo != null) lower[i] = higher[map.get(lo)];
            if(hi != null) higher[i] = lower[map.get(hi)];

            if(higher[i]) res++;

            map.put(arr[i], i);
        }
        return res;
    }
}
